tester/assembly/unittest.ts

```diff
@@ -27,39 +28,58 @@ class TestNode {
   success: bool = false;
   constructor(
     public name: string,
-    public callback: () => void,
+    public callback: () => i32,
   ) { }
 
-  evaluate(ctx: TestNodeReporterContext): void {
+  evaluate(ctx: TestNodeReporterContext): i32 {
     if (this != root) {
       ctx.indent += 2;
-      if (this.group) write(' '.repeat(ctx.indent) + 'Group: ' + this.name + '\n');
+      if (this.group) write(' '.repeat(ctx.indent) + 'Set: ' + this.name + '\n');
       else write(' '.repeat(ctx.indent) + 'Test: ' + this.name + '\n');
     }
 
     const parent = current;
     current = this;
-    this.callback();
+
+    let returnCode = this.callback();
+
+    if (returnCode == -1) {
+      return -1;
+    }
 
     // once the test is run, children are determined, evaluate them
     const children = this.children;
     const childrenLength = children.length;
     for (let i = 0; i < childrenLength; i++) {
       const child = unchecked(children[i]);
-      child.evaluate(ctx);
+
+      const childResultCode = child.evaluate(ctx);
+
+      if (childResultCode == -1) {
+        returnCode = -1;
+        break;
+      }
+
+      if (childResultCode == 0) {
+        returnCode = 0;
+      }
```

Comment:
returnCode = child.evaluate(ctx);
    if (returnCode < 0) {
        break;
      }

Reply:
I don't think we can do that because returnCode need to be absorbent : if one of the test has failed, returnCode must never returned to a success value (1).

Using the code you propose, if the last test is a success, even if previous tests failed, the return code will be a success.

Now after explaining that I guess that I need to clarify the code by using constants instead of number(-1, 0, 1)...

Reply:
@peterjah If you can have a look at my last commit to tell me if the code is clearer.

Reply:
much better without magic numbers :)